### PR TITLE
add user/pass based Auth

### PR DESCRIPTION
**Pull request by CFMS4-Ali**
Added user password authentication for kotlin. Based on the new Memphis changes

### CODE REVIEW

**Comment by yanivbh1** [comment]
Thank you very much for the contribution @CFMS4-Ali!
Adding @adrianNEMO and @rnowling-memphis to review and approve the PR.

File: src/main/kotlin/dev/memphis/sdk/Memphis.kt [hunk 1]
```diff
@@ -212,15 +230,32 @@ class Memphis private constructor(
         var connectionTimeout = 15.seconds;
 
         internal fun build() = Memphis(
-            host, username, connectionToken, port, autoReconnect, maxReconnects, reconnectWait, connectionTimeout
+            host, username, connectionToken ?: "", password ?: "", port, autoReconnect, maxReconnects, reconnectWait, connectionTimeout
```
Comment (rnowling-memphis) [inline src/main/kotlin/dev/memphis/sdk/Memphis.kt:233]: Can you make this take the token or password if you unify the types?

**Review by rnowling-memphis — CHANGES_REQUESTED** [review]
Thank you for this great contribution!  I think my biggest comment is I think that there is opportunity to improve the type safety of this.  In particular, if Password and ConnectionToken can inherit from a common base or be implemented as enums, then a single field can store the authorization value.

Also, please add some documentation to the Password and ConnectionToken types, the methods that take them, and the examples explaining how the user should use them?

If anything is unclear or you disagree with what I'm suggesting, let me know.  :)

Thank you!

File: examples/src/main/kotlin/dev/memphis/Consumer.kt [hunk 2]
```diff
@@ -9,7 +9,7 @@ import kotlinx.coroutines.runBlocking
 
 fun main() {
     runBlocking {
-        val memphis = Memphis.connect("<memphis-host>", "<application type username>", "<broker-token>")
```
[the remark this answers is not in the document]
Comment (CFMS4-Ali) [reply examples/src/main/kotlin/dev/memphis/Consumer.kt:12]: @rnowling-memphis It seems that user/pass Authentication has become the default, it's why I changed the example. We could have both, It's seems straight forward if your using a connectionToken to just replace Memphis.Password with Memphis.ConnectionToken,

File: src/main/kotlin/dev/memphis/sdk/Memphis.kt [hunk 3]
```diff
@@ -67,6 +59,31 @@ class Memphis private constructor(
         brokerConnection.close()
     }
 
+    private fun getTokenBrokerConnection() : Connection {
```
[the remark this answers is not in the document]
Comment (CFMS4-Ali) [reply src/main/kotlin/dev/memphis/sdk/Memphis.kt:62]: @rnowling-memphis good point, I wasn't too happy with that either.

[the remark this answers is not in the document]
Comment (rnowling-memphis) [reply src/main/kotlin/dev/memphis/sdk/Memphis.kt:62]: That makes sense.  Works for me!

File: src/main/kotlin/dev/memphis/sdk/Memphis.kt [hunk 4]
```diff
@@ -212,15 +230,32 @@ class Memphis private constructor(
         var connectionTimeout = 15.seconds;
 
         internal fun build() = Memphis(
-            host, username, connectionToken, port, autoReconnect, maxReconnects, reconnectWait, connectionTimeout
+            host, username, connectionToken ?: "", password ?: "", port, autoReconnect, maxReconnects, reconnectWait, connectionTimeout
         )
     }
+    class Password constructor(
```
[the remark this answers is not in the document]
Comment (CFMS4-Ali) [reply src/main/kotlin/dev/memphis/sdk/Memphis.kt]: @rnowling-memphis  updated please review ( updated the code for all your comments. but ill just reply on one)

File: src/main/kotlin/dev/memphis/sdk/Memphis.kt [hunk 5]
```diff
@@ -43,16 +44,12 @@ class Memphis private constructor(
...
-            .reconnectWait(reconnectWait.toJavaDuration())
-            .maxReconnects(maxReconnects)
-            .let { if (autoReconnect) it else it.noReconnect() }
-            .build()
-            .let { Nats.connect(it) }
+        if (authorizationType is ConnectionToken)
+            buildBrokerConnection().token(authorizationType.connectionToken).build().let { Nats.connect(it) }
+        else if (authorizationType is Password)
+            buildBrokerConnection().userInfo(username, authorizationType.password).build().let { Nats.connect(it) }
+    else
```
Comment (rnowling-memphis) [inline src/main/kotlin/dev/memphis/sdk/Memphis.kt:51]: Indentation error?

**Review by rnowling-memphis — CHANGES_REQUESTED** [review]
Other than a potential indentation error, I would be happy to merge.  Thank you for your contributions!

Reading through your code and the idea of a unified authentication type led me to add the same thing to the Java SDK so your impact was multiplied.  :)

File: src/main/kotlin/dev/memphis/sdk/Memphis.kt (hunk 5 above)
Comment (CFMS4-Ali) [reply src/main/kotlin/dev/memphis/sdk/Memphis.kt]: fixed

**Review by rnowling-memphis — APPROVED** [review]

**Pull request merged**